internal/engine/upper.go

```diff
@@ -224,6 +224,94 @@ func handleBuildStarted(ctx context.Context, state *store.EngineState, action bu
+	// Record the last result
+	status.LastResult = result
+
+	// Remove pending file changes that were consumed by this build.
+	for file, modTime := range status.PendingFileChanges {
+		if store.BeforeOrEqual(modTime, br.StartTime) {
+			delete(status.PendingFileChanges, file)
+		}
+	}
+
+	// Remove pending dependency changes that were consumed by this build.
+	for id, modTime := range status.PendingDependencyChanges {
+		if store.BeforeOrEqual(modTime, br.StartTime) {
+			delete(status.PendingDependencyChanges, id)
+		}
+	}
+
+	// If the whole build succeeded, mark all of the images in the
+	// build as successful.
+	//
+	// This has a minor bug: if one of the images in a set failed, but another one
+	// succeeded, this doesn't mark any of them as successful. We should fix this,
+	// though personally, I think LastSuccessfulResult has problematic semantics
+	// to begin with, and it might make sense to replace it entirely.
+	if isBuildSuccess {
+		status.LastSuccessfulResult = result
+	}
+}
+
+// When a Manifest build finishes, update the BuildStatus for all applicable
+// targets in the engine state.
+func handleBuildResults(engineState *store.EngineState,
+	mt *store.ManifestTarget, br model.BuildRecord, results store.BuildResultSet) {
+	isBuildSuccess := br.Error == nil
+
+	// Update all the build statuses for the current manifest.
+	for _, target := range mt.Manifest.TargetSpecs() {
+		status := mt.State.MutableBuildStatus(target.ID())
+		updateBuildStatus(status, br, results[target.ID()], isBuildSuccess)
+	}
+
+	// Update build statuses for duplicated image targets in other manifests.
+	// This ensures that those images targets aren't rebuilt.
+	engineState.ForEachMutableBuildStatusInResults(results, func(id model.TargetID, result store.BuildResult, mn model.ManifestName, status *store.BuildStatus) {
+		if mt.Manifest.Name == mn {
+			return
+		}
+
+		_, isImageResult := result.(store.ImageBuildResult)
+		if !isImageResult {
+			return
+		}
+
+		updateBuildStatus(status, br, result, isBuildSuccess)
+	})
+
+	// Suppose we built manifestA, which contains imageA depending on imageCommon.
+	//
+	// We also have manifestB, which contains imageB depending on imageCommon.
+	//
+	// We need to mark imageB as dirty, because it was not built in the manifestA
+	// build but its dependency was built.
+	engineState.ForEachMutableBuildStatusDependingOn(results, func(id, reverseDepID model.TargetID, mn model.ManifestName, reverseDepStatus *store.BuildStatus) {
+		if mn == mt.Manifest.Name {
+			// We're only worried about manifests that we didn't build.
+			return
+		}
+
+		_, ok := results[reverseDepID]
+		if ok {
+			// The reverseDep is fresh! Skip it.
+			return
+		}
+
+		_, depIsNew := results[id]
+		if !depIsNew {
+			// This build didn't actually produce a new result. So no need to mark other
+			// manifests dirty.
+			return
+		}
+
+		// This target was not built, but one of its dependencies was
+		// built, so we have to mark it for rebuild.
```

Comment:
(or otherwise clarify the object of the sentence, it's ambiguous :-/ )
```suggestion
		// built, so we have to mark the target for rebuild.
```

Reply:
done